### PR TITLE
SAK-44153 Improvements to Group and Date Released Subpages in Lessons

### PR DESCRIPTION
**Pull request by davidpbauer**
- Updated lesson page so hidden subpages would show "released on" instead of "Not released until" even
after the release date.
- Changed released color to be green. The color of the text will toggle depending on the state of whether
or not the page is visible.
- Links are visible but disabled in the tool menu if they are not released
- Make lessons subpage navigation group aware in the tool menu
- Add option for visible but inactive subpages
- Added release string to date released subpages for student view
- Added checkbox to hide subpage on the main subpage edit screen rather than in gear icon
- Add release date and Hidden properties to those copied in duplication

### CODE REVIEW

File: lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java [hunk 1]
```diff
@@ -1893,7 +1896,18 @@ public String getLessonSubPageJSON(final String userId, final boolean isInstruct
 			fields[i+1] = pageIds.get(i);
 		}
 
-		final LessonsSubNavBuilder lessonsSubNavBuilder = new LessonsSubNavBuilder(siteId, canSeeAll);
+		ArrayList groups = new ArrayList();
```
Comment (adrianfish) [inline lessonbuilder/components/src/java/org/sakaiproject/lessonbuildertool/model/SimplePageToolDaoImpl.java:1899]: Why not type this?
List<String> groups = new ArrayList<>();

Actually:
final List<String> groupIds = siteService.getSite(siteId).getGroupsWithMember(userId).stream().map(Group::getId).collect(Collectors.toList());

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/service/LessonBuilderEntityProducer.java [hunk 2]
```diff
@@ -1120,6 +1122,18 @@ public String merge(String siteId, Element root, String archivePath, String from
 		     String folder = pageElement.getAttribute("folder");
 		     if (StringUtils.isNotEmpty(folder))
 			 page.setFolder(folder);
+		     //get new page's Date Release property
```
Comment (adrianfish) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/service/LessonBuilderEntityProducer.java:1125]: Indentation

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 3]
```diff
@@ -3555,19 +3555,19 @@ public String getReleaseString(SimplePageItem i, Locale locale) {
...
-		     }
-		     return null;
+		 if (page.getReleaseDate() != null ) {
+		     DateFormat df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT, locale);
+		     TimeZone tz = userTimeService.getLocalTimeZone();
+		     df.setTimeZone(tz);
+			 String releaseDate = df.format(page.getReleaseDate());
+		     if(Instant.now().isBefore(page.getReleaseDate().toInstant())){
+				 return messageLocator.getMessage("simplepage.pagenotreleased").replace("{}", releaseDate);
+			 } else{
```
Comment (adrianfish) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:3565]: Indentations gone a bit crazy here :)

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 4]
```diff
@@ -5063,13 +5063,11 @@ public boolean isItemVisible(SimplePageItem item, SimplePage page, boolean testp
...
 		    return true;  // breaks are always visible to all users
 		else if (item.getType() == SimplePageItem.PAGE) {
-		  if (!item.isRequired()) {
 		    SimplePage itemPage = getPage(Long.valueOf(item.getSakaiId()));
 		    if (itemPage.isHidden())
 			return false;
 		    if (itemPage.getReleaseDate() != null && itemPage.getReleaseDate().after(new Date()))
-			return false;
-		  }
+			return true;
```
Comment (adrianfish) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:5070]: Need a opening brace after the if statement, close after. I know lessons does this all over the place, Python style, but still ...

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 5]
```diff
@@ -1735,6 +1722,23 @@ else if (forum.notPublished())
...
+						UIOutput.make(tableRow, "page-button", Boolean.toString("button".equals(i.getFormat())));
+						SimplePage page = simplePageToolDao.getPage(Long.valueOf(i.getSakaiId()));
+						UIOutput.make(tableRow, "page-hidden", Boolean.toString(page.isHidden()));
+						itemGroupString = simplePageBean.getItemGroupString(i, null, true);
+						UIOutput.make(tableRow, "item-groups", itemGroupString);
+						SimplePage sPage = simplePageBean.getPage(Long.parseLong(i.getSakaiId()));
+						if (sPage != null) {
+							Date rDate = sPage.getReleaseDate();
+							String rDateString = "";
+							if (rDate != null)
```
Comment (adrianfish) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java:1737]: Can you wrap some braces around this?

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java [hunk 6]
```diff
@@ -3939,8 +3991,27 @@ protected static boolean makeLink(UIContainer container, String ID, SimplePageIt
...
+							M_locale = new Locale("en", "GB");
+						} else {
+							M_locale = new Locale(langLoc[0], langLoc[1]);
+						}
+					} else {
+						M_locale = new Locale(langLoc[0]);
+					}
+
+					String releaseString = simplePageBean.getReleaseString(i, M_locale);
+		    		link.decorate(new UITooltipDecorator(releaseString));
```
Comment (adrianfish) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/producers/ShowPageProducer.java:4010]: Indentation

File: lessonbuilder/tool/src/webapp/js/show-page.js [hunk 7]
```diff
@@ -1709,6 +1709,14 @@ $(document).ready(function() {
 				$("#item-button").prop("checked", false);
 			    }
 
+			    var pagehidden = row.find(".page-hidden").text();
+			    if(pagehidden === "true") {
+			    	$("#hide2").prop("checked", true);
+			    	$("#hide2").attr("defaultChecked", true);
```
Comment (adrianfish) [inline lessonbuilder/tool/src/webapp/js/show-page.js:1715]: You can chain these calls.

File: lessonbuilder/tool/src/webapp/js/show-page.js [hunk 8]
```diff
@@ -1709,6 +1709,14 @@ $(document).ready(function() {
...
+				}else {
+			    	$("#hide2").prop("checked", false);
```
Comment (adrianfish) [inline lessonbuilder/tool/src/webapp/js/show-page.js:1717]: Indentation

**Review by adrianfish — APPROVED** [review]
I'm approving this, but there are a few suggested changes. I tested it out locally, and it all seems to work and is, imo, not controversial.

**Comment by davidpbauer** [comment]
Thanks for reviewing, @adrianfish. I've pushed another commit that addresses your comments by cleaning up the old Lessons code.

**Pull request merged**